qt/widgets/common/src/BasicFitOptionsBrowser.cpp

```diff
@@ -0,0 +1,229 @@
+  m_browser->addProperty(m_minimizer);
+
+  addProperty("Minimizer", m_minimizer,
+              &BasicFitOptionsBrowser::getStringEnumProperty,
+              &BasicFitOptionsBrowser::setStringEnumProperty);
+}
+
+void BasicFitOptionsBrowser::createCostFunctionProperty() {
+  m_costFunction = m_enumManager->addProperty("Cost Function");
+
+  m_enumManager->setEnumNames(m_costFunction, costFunctions());
+  m_browser->addProperty(m_costFunction);
+
+  addProperty("Cost Function", m_costFunction,
+              &BasicFitOptionsBrowser::getStringEnumProperty,
+              &BasicFitOptionsBrowser::setStringEnumProperty);
+}
+
+void BasicFitOptionsBrowser::createEvaluationTypeProperty() {
+  m_evaluationType = m_enumManager->addProperty("Evaluation Type");
+
+  m_enumManager->setEnumNames(m_evaluationType, EVALUATION_TYPES);
+  m_browser->addProperty(m_evaluationType);
+
+  addProperty("Evaluation Type", m_evaluationType,
+              &BasicFitOptionsBrowser::getStringEnumProperty,
+              &BasicFitOptionsBrowser::setStringEnumProperty);
+}
+
+void BasicFitOptionsBrowser::addProperty(std::string const &name,
+                                         QtProperty *prop,
+                                         PropertyGetter getter,
+                                         PropertySetter setter) {
+  m_propertyNameMap[name] = prop;
+  m_getters[prop] = getter;
+  m_setters[prop] = setter;
+}
+
+void BasicFitOptionsBrowser::setProperty(std::string const &name,
+                                         std::string const &value) {
+  auto const prop = getQtPropertyFor(name);
+  auto const f = m_setters[prop];
+  (this->*f)(prop, value);
+}
+
+std::string BasicFitOptionsBrowser::getProperty(std::string const &name) const {
+  auto const prop = getQtPropertyFor(name);
+  auto const f = m_getters[prop];
+  return (this->*f)(prop);
+}
+
+QtProperty *
+BasicFitOptionsBrowser::getQtPropertyFor(std::string const &name) const {
+  if (!m_propertyNameMap.contains(name)) {
+    throw std::runtime_error("Property " + name +
+                             " isn't supported by the browser.");
+  }
+  return m_propertyNameMap[name];
+}
+
+void BasicFitOptionsBrowser::enumChanged(QtProperty *prop) {
+  if (prop == m_fittingMode) {
+    emit fittingModeChanged(getFittingMode());
+  }
+}
+
+void BasicFitOptionsBrowser::setIntProperty(QtProperty *prop,
+                                            std::string const &value) {
+  m_intManager->setValue(prop, QString::fromStdString(value).toInt());
+}
+
+std::string BasicFitOptionsBrowser::getIntProperty(QtProperty *prop) const {
+  return QString::number(m_intManager->value(prop)).toStdString();
+}
+
+void BasicFitOptionsBrowser::setStringEnumProperty(QtProperty *prop,
+                                                   std::string const &value) {
+  auto const i =
+      m_enumManager->enumNames(prop).indexOf(QString::fromStdString(value));
+  if (i >= 0)
```

Comment:
Is it worth throwing an invalid argument exception like you do lower down if the property isn't found? Not 100% sure but maybe if someone added a new property and forgot to do m_enumManager->addProperty(XXX) this exception might be useful